### PR TITLE
Refactor/pro 4320/secondary interface

### PR DESCRIPTION
**Pull request by patrickvacek**
This is still a bit of a work-in-progress (as evidenced by all the TODOs) but this at least compiles and passes all tests (thanks @OYTIS!). It certainly covers redoing the interface, but the actual implementation still has some details to work out.

### CODE REVIEW

File: src/sotauptaneclient.cc [hunk 1]
```diff
@@ -1,17 +1,33 @@
...
+#include "uptane/secondaryconfig.h"
+#include "uptane/secondaryfactory.h"
 #include "utils.h"
 
 SotaUptaneClient::SotaUptaneClient(const Config &config_in, event::Channel *events_channel_in, Uptane::Repository &repo)
-    : config(config_in), events_channel(events_channel_in), uptane_repo(repo), last_targets_version(-1) {}
+    : config(config_in), events_channel(events_channel_in), uptane_repo(repo), last_targets_version(-1) {
+  std::vector<Uptane::SecondaryConfig>::iterator it;
+  for (it = config.uptane.secondary_configs.begin(); it != config.uptane.secondary_configs.end(); ++it) {
+    std::map<std::string, boost::shared_ptr<Uptane::SecondaryInterface> >::const_iterator map_it =
```
Comment (patriotyk) [inline src/sotauptaneclient.cc]: We can use std:set for unique elements

File: src/uptane/legacysecondary.cc [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#include "crypto.h"
+#include "logger.h"
+
+namespace Uptane {
+LegacySecondary::LegacySecondary(const SecondaryConfig& sconfig_in) : ManagedSecondary(sconfig_in) {
+  boost::filesystem::create_directories(sconfig.firmware_path.parent_path());
+}
+
+bool LegacySecondary::storeFirmware(const std::string& target_name, const std::string& content) {
```
Comment (patriotyk) [inline src/uptane/legacysecondary.cc]: Firmware might be very big and store it in memory may consume all of free memory.

File: src/uptane/managedsecondary.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  std::string detected_attack;
+  std::string expected_target_name;
+  std::vector<Hash> expected_target_hashes;
+  int64_t expected_target_length;
+
+  MetaPack current_meta;
+
+  virtual bool storeFirmware(const std::string& target_name, const std::string& content) = 0;
+  // TODO: bool getFirmwareInfo(std::string* target_name, size_t& target_len, std::vector<Uptane::Hash>* hashes) {};
+  virtual bool getFirmwareInfo(std::string* target_name, size_t& target_len, std::string* sha256hash) {
```
Comment (patriotyk) [inline src/uptane/managedsecondary.h]: Why it is not a pure virtual?

File: src/uptane/virtualsecondary.cc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  Utils::writeFile(sconfig.firmware_path.string(), content);
+  return true;
+}
+
+bool VirtualSecondary::getFirmwareInfo(std::string* target_name, size_t& target_len, std::string* sha256hash) {
+  if (!boost::filesystem::exists(sconfig.target_name_path) || !boost::filesystem::exists(sconfig.firmware_path)) {
+    *target_name = std::string("noimage");
+    target_len = 0;
+    *sha256hash = std::string("");
+    return true;
```
Comment (patriotyk) [inline src/uptane/virtualsecondary.cc]: maybe return false?

File: src/sotauptaneclient.cc (hunk 1 above)
Comment (OYTIS) [reply src/sotauptaneclient.cc]: The point here is to be able to quickly find a secondary by its serial number.

File: src/uptane/virtualsecondary.cc (hunk 4 above)
Comment (OYTIS) [reply src/uptane/virtualsecondary.cc]: No, false is reserved for connection failures, which don't happen in virtual secondaries.

File: src/uptane/managedsecondary.h (hunk 3 above)
Comment (patrickvacek) [reply src/uptane/managedsecondary.h]: You are right, probably should be a pure virtual function. That was leftover cruft.

**Comment by cajun-rat** [comment]
When building this in meta-updater, the device doesn't register correctly in the UI.  I'll do some investigation...

**Comment by patrickvacek** [comment]
@cajun-rat Frankly I'm not surprised that there are still some problems. Although do you mean even the primary doesn't show up?

**Comment by cajun-rat** [comment]
Yeah, no primary and no packages

**Comment by cajun-rat** [comment]
Are you able to make this work against ATS Garage? I still haven't got a device to provision correctly in the GUI

**Comment by patrickvacek** [comment]
@cajun-rat No, still working on it. I think I found the problem but still sorting out the details. But I think I resolved the one issue @patriotyk brought up that @OYTIS didn't already answer.

**Comment by cajun-rat** [comment]
Tested in with an update to Qemu inside meta-updater

**Pull request merged**